reactive/calico.py

```diff
@@ -259,6 +321,72 @@ def ready():
         status_set('active', 'Calico is active')
 
 
+@when('config.changed.registry-credentials')
+def registry_credentials_changed():
+    status_set('maintenance', 'Applying registry credentials')
+    encoded_creds = hookenv.config('registry-credentials')
+
+    if not encoded_creds:
+        log('no registry-credentials, skipping docker config')
+        return
+
+    creds = b64decode(encoded_creds).decode('utf-8')
+    config_dir = '/root/.docker'
+    config_path = config_dir + '/config.json'
+    os.makedirs(config_dir, exist_ok=True)
+    with open(config_path, 'w') as f:
+        f.write(creds)
+    remove_state('calico.npc.deployed')
+
+
+@when('config.changed.registry')
+def registry_changed():
+    remove_state('calico.service.installed')
+    remove_state('calico.npc.deployed')
+
+
+@when('calico.npc.deployed', 'kube-api-endpoint.available')
+def watch_for_api_endpoint_changes():
+    apiserver_ips = get_apiserver_ips()
+    old_apiserver_ips = db.get('tigera.apiserver_ips_used')
+    if apiserver_ips != old_apiserver_ips:
+        log('apiserver endpoints changed, preparing to reapply templates')
+        remove_state('calico.npc.deployed')
+
+
+def get_apiserver_ips():
```

Comment:
I would prefer to see helper functions like this live in `lib/charm/layer/tigera_secure_ee.py`.  Of course, I'd also generally like to see us start breaking handlers up into separate files based on logical groupings, so this isn't really a blocker for now.

Reply:
I'll open a new PR to move these to `lib/charm/layer/tigera_secure_ee.py`. Thanks :+1:

Reply:
On second thought, I'm gonna have to punt on this since I'm falling behind on the timeline for this charm. Sorry.